src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php

```diff
@@ -0,0 +1,465 @@
+
+                $route->setDefaults(array_merge($this->defaults, $route->getDefaults()));
+                $route->setOptions(array_merge($this->options, $route->getOptions()));
+
+                // we're extra careful here to avoid re-setting deprecated _method and _scheme
+                foreach ($this->requirements as $key => $val) {
+                    if (!$route->hasRequirement($key)) {
+                        $route->setRequirement($key, $val);
+                    }
+                }
+
+                if (null !== $this->prefix) {
+                    $route->setPath('/'.$this->prefix.$route->getPath());
+                }
+
+                if (!$route->getHost()) {
+                    $route->setHost($this->host);
+                }
+
+                if (!$route->getCondition()) {
+                    $route->setCondition($this->condition);
+                }
+
+                if (!$route->getSchemes()) {
+                    $route->setSchemes($this->schemes);
+                }
+
+                if (!$route->getMethods()) {
+                    $route->setMethods($this->methods);
+                }
+
+                $routeCollection->add($name, $route);
+            } else {
+                /* @var self $route */
+
+                $subCollection = $route->build();
+                $subCollection->addPrefix($this->prefix);
+
+                $routeCollection->addCollection($subCollection);
+            }
+        }
+
+        foreach ($this->resources as $resource) {
+            $routeCollection->addResource($resource);
+        }
+
+        return $routeCollection;
+    }
+
+    /**
+     * Attempts to safely prefix controllers with the controller class if necessary.
+     *
+     * @param Route $route
+     */
+    private function ensureRouteController(Route $route)
+    {
+        // only do work if there is a controller class set
+        if (!$this->controllerClass) {
+            return;
+        }
+
+        $controller = $route->getDefault('_controller');
+
+        // only apply controller class to a (non-empty) string
+        if (!is_string($controller) || !$controller) {
+            return;
+        }
+
+        // is the controller already a callable function/class?
+        if (method_exists($controller, '__invoke') || function_exists($controller)) {
+            return;
+        }
+
+        // is this already a controller format (a:b:c, or a:b, or a::b)?
+        if (false !== strpos($controller, ':')) {
+            return;
+        }
+
+        $controller = sprintf('%s::%s', $this->controllerClass, $controller);
+        $route->setDefault('_controller', $controller);
+    }
+
+    /**
+     * Generates a route name based on details of this route.
+     *
+     * @return string
+     */
+    private function generateRouteName(Route $route)
+    {
+        $methods = implode('_', $route->getMethods()).'_';
+
+        $routeName = $methods.$route->getPath();
```

Comment:
The prefix is not yet part of `getPath()` here as it is added later in the process, so we can have collisions here. You should either pass the prefix and add it here manually, or call `generateRouteName()` later in `build()`.